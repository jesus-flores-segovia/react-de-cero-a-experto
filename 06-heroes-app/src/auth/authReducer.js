import { types } from "../types/types";

export const authReducer = (state = {}, action) => {

    switch (action.type) {
        case types.login:
            return {
                ...state,
                name: action.payload,
                logged: true
            }
        case types.logout:
            return {
                ...state,
                name: "",
                logged: false
            }
    
        default:
            return state;
    }
}