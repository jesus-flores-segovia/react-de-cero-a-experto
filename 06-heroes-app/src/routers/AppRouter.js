import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route
} from "react-router-dom";
import { DcScreen } from "../components/dc/DcScreen";
import { LoginScreen } from "../components/login/LoginScreen";
import { MarvelScreen } from "../components/marvel/MarvelScreen";
import { Navbar } from "../components/ui/NavBar";

export const AppRouter = () => {
    return (
        <Router>
          <div>
            <Navbar/>
                <Switch>
                    <Route exact path="/login" component={LoginScreen}/>
                    <Route exact path="/marvel" component={MarvelScreen}/>
                    <Route exact path="/dc" component={DcScreen}/>
                </Switch>
          </div>
        </Router>
      );
    }
