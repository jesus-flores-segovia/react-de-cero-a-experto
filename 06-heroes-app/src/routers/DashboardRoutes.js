import React from 'react';
import {
    Switch,
    Route,
    Redirect
  } from "react-router-dom";
import { HeroScreen } from '../components/heroes/HeroScreen';
import { DcScreen } from "../components/dc/DcScreen";
import { MarvelScreen } from "../components/marvel/MarvelScreen";
import { Navbar } from '../components/ui/NavBar';

export const DashboardRoutes = () => {
    return (
        <>
            <Navbar/>
            <div className="container mt-5">
                <Switch>
                    <Route exact path="/marvel" component={MarvelScreen}/>
                    <Route exact path="/hero/:heroId" component={HeroScreen}/>
                    <Route exact path="/dc" component={DcScreen}/>

                    <Redirect to="/marvel"/>
                </Switch>
            </div>   
        </>
    )
}
