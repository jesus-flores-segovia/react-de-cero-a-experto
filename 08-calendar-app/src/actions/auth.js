import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { calendarLogoutCleanEvents } from "./calendar";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth",
        { email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { email, password } = body.errors;
          const message = email ? email.msg : password.msg;
          Swal.fire("Cannot login", message, "error");
          return;
        }

        Swal.fire("Cannot login", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth/register",
        { name, email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { name, email, password } = body.errors;
          const message = name ? name.msg : email ? email.msg : password.msg;
          Swal.fire("Cannot register", message, "error");
          return;
        }

        Swal.fire("Cannot register", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startCheckingLoginState = () => {
  return async (dispatch) => {
    try {
      dispatch(checkLoginState({ isCheckingLoginState: true }));
      const response = await fetchWithToken("auth/newToken");
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      }
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    } catch (error) {
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    }
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(calendarLogoutCleanEvents());
    dispatch(logout());
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const checkLoginState = (isCheckingLoginState) => ({
  type: types.authCheckLoginState,
  payload: isCheckingLoginState,
});

const logout = () => ({
  type: types.authLogout,
});
