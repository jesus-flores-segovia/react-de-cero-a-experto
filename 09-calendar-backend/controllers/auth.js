const { response } = require("express");
const { validationResult } = require("express-validator");

const register = (req, res = response) => {
  const { name, email, password } = req.body;
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).send({ ok: false, errors: errors.mapped() });
  }

  res.status(201).send({ ok: true, msg: "register", name, email, password });
};

const login = (req, res = response) => {
  const { email, password } = req.body;
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).send({ ok: false, errors: errors.mapped() });
  }

  res.status(201).send({ ok: true, msg: "login", email, password });
};

const newToken = (req, res = response) => {
  res.send({ ok: true, msg: "new token" });
};

module.exports = {
  register,
  login,
  newToken,
};
