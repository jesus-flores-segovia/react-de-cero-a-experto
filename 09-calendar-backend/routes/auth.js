/* 
    Routing for the users auth
    url: host + /api/auth
*/

const express = require("express");
const { check } = require("express-validator");
const router = express.Router();

const { register, login, newToken } = require("../controllers/auth");

router.post(
  "/register",
  [
    check("name", "Name is mandatory").not().isEmpty(),
    check("email", "Email is mandatory").isEmail(),
    check("password", "Password must be at least of 6 characters").isLength({
      min: 6,
    }),
  ],
  register
);

router.post(
  "/",
  [
    check("email", "Email is mandatory").isEmail(),
    check("password", "Password must be at least of 6 characters").isLength({
      min: 6,
    }),
  ],
  login
);

router.get("/newToken", newToken);

module.exports = router;
