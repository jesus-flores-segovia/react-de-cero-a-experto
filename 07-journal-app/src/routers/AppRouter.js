import { onAuthStateChanged } from 'firebase/auth'
import React, { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { BrowserRouter, Route, Routes, Navigate } from 'react-router-dom'
import { login } from '../actions/auth'
import { startLoadingNotes } from '../actions/notes'

import { JournalScreen } from '../components/journal/JournalScreen'
import { auth } from '../firebase/firebaseConfig'
import { AuthRouter } from './AuthRouter'
import { PrivateRoute } from './PrivateRoute'
import { PublicRoute } from './PublicRoute'

export const AppRouter = () => {

    const dispatch = useDispatch();

    const [checking, setChecking] = useState(true);
    const [isLoggedIn, setIsLoggedIn] = useState(false)

    useEffect(() => {
        onAuthStateChanged(auth, (user) => {
            if(user?.uid) {
                dispatch(login(user.uid, user.displayName));
                setIsLoggedIn(true);

                dispatch(startLoadingNotes());
            } else{
                setIsLoggedIn(false);
            }

            setChecking(false);
        })
    }, [dispatch, setChecking, setIsLoggedIn])

    if(checking) {
        return (
            <h1>Wait...</h1>
        )
    }

    return (
        <BrowserRouter>
            <Routes>
                <Route path="/" 
                    element={
                        <PrivateRoute isLoggedIn={isLoggedIn}>
                            <JournalScreen/>
                        </PrivateRoute>
                        }/>
                <Route path="auth/*" 
                    element={
                        <PublicRoute isLoggedIn={isLoggedIn}>
                            <AuthRouter/>
                        </PublicRoute>
                        }/>
                <Route 
                    path="*"
                    element={<Navigate to="/auth/*" replace={true} />}
                />
            </Routes>
        </BrowserRouter> 
    )
}
