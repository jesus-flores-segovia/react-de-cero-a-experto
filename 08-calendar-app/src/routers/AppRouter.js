import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Navigate, Route, Routes } from "react-router-dom";
import { startCheckingLoginState } from "../actions/auth";
import { LoginScreen } from "../components/auth/LoginScreen";
import { CalendarScreen } from "../components/calendar/CalendarScreen";
import { PrivateRoute } from "./PrivateRoute";
import { PublicRoute } from "./PublicRoute";

export const AppRouter = () => {
  const dispatch = useDispatch();

  const { isLoggedIn, isCheckingLoginState } = useSelector(
    (state) => state.auth
  );

  useEffect(() => {
    dispatch(startCheckingLoginState());
  }, [dispatch]);

  if (isCheckingLoginState) {
    return <h1>Checking</h1>;
  }

  return (
    <Routes>
      <Route
        path="/"
        element={
          <PrivateRoute isLoggedIn={isLoggedIn} redirectPath={"login"}>
            <CalendarScreen />
          </PrivateRoute>
        }
      />
      <Route
        path="login"
        element={
          <PublicRoute isLoggedIn={isLoggedIn} redirectPath={"/"}>
            <LoginScreen />
          </PublicRoute>
        }
      />
      <Route path="login" element={<LoginScreen />} />
      <Route path="*" element={<Navigate to="/" />} />
    </Routes>
  );
};
