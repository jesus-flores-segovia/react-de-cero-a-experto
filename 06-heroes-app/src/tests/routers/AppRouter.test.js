import { mount } from "enzyme"
import { AuthContext } from "../../auth/AuthContext"
import { AppRouter } from "../../routers/AppRouter"

describe("Tests inside file 'AppRouter.test.js'", () => {

    test("Must show the login if the user isn't authenticated", () => {
        
        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged: false
            }
        }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <AppRouter/>
            </AuthContext.Provider>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("h1").text()).toBe("Login");
    })

    test("Must show the marvel page if the user is authenticated", () => {

        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged: true,
                name:"Jes√∫s"
            }
        }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <AppRouter/>
            </AuthContext.Provider>
        )
        
        expect(wrapper.find(".navbar").exists()).toBeTruthy();
    })
})