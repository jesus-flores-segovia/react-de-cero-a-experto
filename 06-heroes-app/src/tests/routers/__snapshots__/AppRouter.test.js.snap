// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'AppRouter.test.js' Must show the login if the user isn't authenticated 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
